File: LeetCode/211. Design Add and Search Words Data Structure.js
// https://leetcode.com/problems/design-add-and-search-words-data-structure/description/

// В Питон более быстрое решение

class WordDictionary {
  constructor(value = '.') {
    this.value = value;

    this.children = {};
    this.childrenAsArray = [];
    this.isEnd = false;
  }

  addWord(word, index = 0, currentNode = this) {
    const node = currentNode.children[word[index]] ?? new WordDictionary(word[index]);

    if (!currentNode.children[word[index]]) {
      currentNode.children[word[index]] = node;
      currentNode.childrenAsArray.push(node);
    }

    index++;

    if (index < word.length) {
      this.addWord(word, index, node);
    } else {
      node.isEnd = true;
    }
  }

  search(word, index = 0, currentNode = this) {
    if (index === word.length) {
      return currentNode.isEnd;
    }

    const char = word[index];

    if (char === '.') {
      for (const child of currentNode.childrenAsArray) {
        const result = this.search(word, index + 1, child);

        if (result) {
          return true;
        }
      }

      return false;
    }

    if (!(char in currentNode.children)) {
      return false;
    }

    return this.search(word, index + 1, currentNode.children[char]);
  }
}

File: LeetCode/211. Design Add and Search Words Data Structure.py
# https://leetcode.com/problems/design-add-and-search-words-data-structure/description/


class WordDictionary:

    def __init__(self, value: str = "."):
        self.value = value
        self.children: dict[str, WordDictionary] = {}
        self.is_end_of_word = False

    def addWord(self, word: str) -> None:
        current = self

        for char in word:
            if char not in current.children:
                current.children[char] = WordDictionary(char)

            current = current.children[char]

        current.is_end_of_word = True

    def search(self, word: str) -> bool:
        def traverse(index: int, current: WordDictionary) -> bool:
            if index == len(word):
                return current.is_end_of_word

            if word[index] == ".":
                for char in current.children:
                    if traverse(index + 1, current.children[char]):
                        return True

            if word[index] not in current.children:
                return False

            return traverse(index + 1, current.children[word[index]])

        return traverse(0, self)
